### PR TITLE
Add configurable parity-fixing strategy for MST joiner

### LINKED ISSUE #89: Evaluate spanning tree objectives: MST vs. minimax vs. alternatives
## Summary

Evaluate whether the minimum spanning tree (MST) is the right objective for connecting disconnected contours, or whether an alternative spanning tree formulation would produce better results for sand tables and pen plotters.

## Background

The current `mst_join` algorithm minimizes *total* new connecting segment length (MST objective). This is the sum of all connection edge weights. The question is whether this is the best objective for the use case.

### MST (current)
- **Objective:** Minimize total edge weight
- **Property:** Also minimizes the maximum edge weight for Euclidean metrics (bottleneck spanning tree theorem)
- **Behavior:** May accept one moderately long edge if it saves many shorter ones elsewhere
- **Artifacts:** Total new visible line length is minimized

### Minimax / Bottleneck spanning tree
- **Objective:** Minimize the longest edge
- **Property:** For Euclidean metrics, the MST is already a bottleneck spanning tree — so this produces the same result
- **Note:** This means the MST is already optimal for "worst single connection" quality

### Degree-constrained spanning tree
- **Objective:** MST with a maximum vertex degree constraint
- **Motivation:** The Euler path phase must retrace edges to fix odd-degree vertices. A degree-constrained tree could reduce retracing by limiting how many connections meet at a single point
- **Trade-off:** May increase total edge weight

### k-opt local search on MST
- **Objective:** Start with MST, then apply local edge swaps that reduce some secondary metric (e.g., total retrace distance after parity fixing)
- **Motivation:** The MST minimizes connection length but doesn't account for the downstream Euler path quality

### Other formulations
- Steiner tree (adds intermediate points) — probably overkill
- Traveling salesman on contour endpoints — different problem but related
- Minimum weight Euler augmentation directly (skip the MST + parity-fix decomposition)

## Evaluation criteria

1. **Visible artifact quality** — how much new visible line is drawn?
2. **Retrace distance** — how much invisible retracing is needed?
3. **Total path length** — sum of visible + invisible travel
4. **Worst-case single connection** — longest individual connecting segment
5. **Computational cost** — acceptable at typical input sizes

## Approach

Start with a benchmark suite of representative images and measure all five criteria under each formulation. The cherry_blossoms test image is a good starting point.

### PR DESCRIPTION
**Pull request by altendky**
## Summary

- Fix two bugs in the Hierholzer Euler path algorithm that caused long diagonal artifacts (up to 389px) in SVG output
- Add configurable parity-fixing strategy (Greedy/Optimal) for the MST joiner
- Add diagnostic infrastructure for MST edge analysis and segment visualization

### Self-loops in `build_graph` (root cause)
When two MST edge endpoints snapped to the same graph node, `graph.add_edge(na, nb, 0.0)` created a self-loop. Petgraph's `edges(n).count()` counts self-loops once but they contribute 2 to degree, causing an **odd** number of odd-degree vertices (violating the handshaking lemma). This prevented `fix_parity` from reducing to exactly 0 or 2 odd vertices.

**Fix:** Added `na != nb` guards to all `graph.add_edge()` calls in `build_graph`.

### Hierholzer limited to exactly 2 odd vertices
The virtual-edge approach for converting an Euler path into an Euler circuit only activated when `odd.len() == 2`. When the self-loop bug left 1 odd vertex, Hierholzer ran without the fix and produced invalid transitions (the visible diagonal lines).

**Fix:** Generalized to handle any number of odd-degree vertices by pairing them and adding virtual edges for each pair, then removing all virtual-edge transitions after circuit construction.

### Result
| Config | Before | After |
|--------|--------|-------|
| Default (mask=0.75, k=20) | 389px, 369px diagonal artifacts | Longest: 87px (legitimate MST edge) |
| Custom (mask=0.6, k=200) | 369px diagonal | Longest: 87px |

## New features

- **`ParityStrategy` enum** (Greedy/Optimal) — configurable parity-fixing algorithm, plumbed through config, UI, CLI
- **`MstEdgeInfo`** struct with per-MST-edge diagnostic data
- **`to_diagnostic_svg()`** — SVG export with MST edges highlighted in red
- **`to_segment_diagnostic_svg()`** — SVG export highlighting top N longest segments in distinct colors with a legend
- **`format_metrics`** — longest-edge info in diagnostics report
- **Benchmark test** comparing Greedy vs Optimal parity strategies
- **Segment diagnostic test** with default config

## Related issues

- Closes #89 — Evaluate spanning tree objectives
- #137 — Revisit Optimal parity strategy (filed: currently equivalent to Greedy)
- #138 — UI diagnostic overlay toggle (filed for future)

### CODE REVIEW

File: crates/mujou-pipeline/src/mst_join.rs [hunk 1]
```diff
@@ -723,33 +787,314 @@ fn fix_parity(
...
+    while mask != 0 {
+        let remaining = mask;
+        let i = remaining.trailing_zeros() as usize;
+        let j = choice[mask] as usize;
+        pairs.push((odd[i], odd[j]));
+        mask ^= (1 << i) | (1 << j);
+    }
+
+    Ok(pairs)
+}
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-pipeline/src/mst_join.rs]: Good catch — the backtracking was indeed inconsistent with the forward pass. Switched to "to-match" mask semantics so `i = mask.trailing_zeros()` is the same vertex in both directions. Addressed in [`7af69ff`](https://github.com/altendky/mujou/commit/7af69ff206be38c4df28fd3d93a59aa271363f9d).

File: crates/mujou-export/tests/cherry_blossoms.rs [hunk 2]
```diff
@@ -95,3 +95,502 @@ fn cherry_blossoms_default_config() {
...
+    struct StdClock;
+    impl mujou_pipeline::diagnostics::Clock for StdClock {
+        type Instant = Instant;
+        fn now(&self) -> Instant {
+            Instant::now()
+        }
+        fn elapsed(&self, since: &Instant) -> Duration {
+            since.elapsed()
+        }
+    }
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-export/tests/cherry_blossoms.rs]: Moved `StdClock` to a single module-level definition shared by all three tests. Addressed in [`d8f7590`](https://github.com/altendky/mujou/commit/d8f75907701356765d332beddbe8ad8f26d7c937).

File: crates/mujou-export/tests/cherry_blossoms.rs [hunk 3]
```diff
@@ -95,3 +95,502 @@ fn cherry_blossoms_default_config() {
...
+    // Write diagnostic SVG with highlighted MST edges.
+    let diagnostic_svg = mujou_export::to_diagnostic_svg(
+        std::slice::from_ref(&result.joined),
+        result.dimensions,
+        &mujou_export::SvgMetadata {
+            title: Some("cherry-blossoms (MST edge diagnostics)"),
+            description: Some("Red lines = MST connecting edges"),
+        },
+        &quality.mst_edge_details,
+    );
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-export/tests/cherry_blossoms.rs]: The `config_json` field was added on `main` after this branch diverged. Merged main and added `config_json: None` to both test initializers. Addressed in [`c564d7e`](https://github.com/altendky/mujou/commit/c564d7e8c421ec75126889e25b03028f610a4554).

File: crates/mujou-pipeline/src/mst_join.rs [hunk 4]
```diff
@@ -881,6 +1304,81 @@ fn hierholzer(graph: &UnGraph<(), f64>) -> Vec<NodeIndex> {
...
+            // Build the result: wrap-around fragment first, then interiors.
+            let mut result = Vec::with_capacity(path.len() - virtual_positions.len());
+            result.extend_from_slice(tail);
+            result.extend_from_slice(head);
+            for frag in fragments {
+                result.extend_from_slice(frag);
+            }
+            return result;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-pipeline/src/mst_join.rs]: Added a comment explaining this branch shouldn't be reached after `fix_parity`, plus a `debug_assert!` that verifies adjacency of every consecutive pair in the assembled path. If discontinuities are detected in debug builds, the assertion fires with the virtual edge positions. Addressed in [`c4920b3`](https://github.com/altendky/mujou/commit/c4920b3aa892c17a7430a0c839fefded1ec21e6d).

File: crates/mujou-export/src/svg.rs [hunk 5]
```diff
@@ -175,6 +282,256 @@ pub fn to_svg(
...
+            );
+        }
+        let _ = writeln!(out, "  </g>");
+    }
+
+    // Closing tag
+    let _ = writeln!(out, "</svg>");
+
+    out
+}
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-export/src/svg.rs]: Replaced the hardcoded `width="340"` with a computed width based on the longest label string. Uses a monospace character-width heuristic (~7.2px/char at font-size 12) plus padding for the color swatch and margins. Addressed in [`8c0c280`](https://github.com/altendky/mujou/commit/8c0c280a01d6e93fdf2a313b68c07e55ab498651).

File: crates/mujou-export/tests/cherry_blossoms.rs [hunk 6]
```diff
@@ -95,3 +109,475 @@ fn cherry_blossoms_default_config() {
...
+    let mst_set: std::collections::HashSet<((i64, i64), (i64, i64))> = quality
+        .mst_edge_details
+        .iter()
+        .flat_map(|e| {
+            let a = ((e.point_a.0 * 10.0) as i64, (e.point_a.1 * 10.0) as i64);
+            let b = ((e.point_b.0 * 10.0) as i64, (e.point_b.1 * 10.0) as i64);
+            // Both directions
+            vec![(a, b), (b, a)]
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-export/tests/cherry_blossoms.rs]: Switched from truncating `as i64` to `(coord * 10.0).round() as i64` for both the HashSet construction and the lookup side, so boundary values map consistently. Addressed in [`9a311d7`](https://github.com/altendky/mujou/commit/9a311d7e40b11fdfef67c859458986b221364970).

File: crates/mujou-pipeline/src/mst_join.rs [hunk 7]
```diff
@@ -723,33 +817,313 @@ fn fix_parity(
...
+        let euclidean_cost = matching_graph_cost(&euclidean_pairs, odd, &dist_matrix);
+        let graph_dist_cost = matching_graph_cost(&graph_dist_pairs, odd, &dist_matrix);
+
+        Ok(if euclidean_cost <= graph_dist_cost {
+            euclidean_pairs
+        } else {
+            graph_dist_pairs
+        })
+    }
+}
```
[the remark this answers is not in the document]
Comment (altendky) [reply crates/mujou-pipeline/src/mst_join.rs]: The distance matrix is used in both the DP path and the heuristic path (`greedy_graph_distance_matching` + `matching_graph_cost`), so the computation isn't redundant. Added a comment documenting the O(n×(V+E)logV) cost and referencing issue #137 for potential sparse-distance optimization. Addressed in [`ddd58c6`](https://github.com/altendky/mujou/commit/ddd58c6449867695957885bc53f50a64393a0990).
